packages/react/runtime/src/snapshot.ts

```diff
@@ -538,18 +547,29 @@ export class SnapshotInstance {
     }
 
     const count = __snapshot_def.slot.length;
-    if (count === 1) {
-      const [, elementIndex] = __snapshot_def.slot[0]!;
+    if (
+      count === 1
+      || (__snapshot_def.isSlotV2 ??= __snapshot_def.slot.every(([type]) =>
+        type === DynamicPartType.SlotV2 || type === DynamicPartType.ListSlotV2
+      ))
+    ) {
+      const [, elementIndex] = __snapshot_def.slot[typeof newNode.__slotIndex === 'number' ? newNode.__slotIndex : 0]!;
       const parent = __elements[elementIndex]!;
       if (shouldRemove) {
         __RemoveElement(parent, newNode.__element_root!);
       }
       if (existingNode) {
-        __InsertElementBefore(
-          parent,
-          newNode.__element_root!,
-          existingNode.__element_root,
-        );
+        if (__snapshot_def.isSlotV2) {
+          if (newNode.__slotIndex! < existingNode.__slotIndex!) {
```

Comment:
It seems this condition will always be `true`. Otherwise this `insertBefore` is redundant and should be eliminated.